scrapy_zyte_api/addon.py

```diff
@@ -0,0 +1,51 @@
+from typing import Optional
+
+from scrapy.settings import BaseSettings
+
+from scrapy_zyte_api import ScrapyZyteAPIDownloaderMiddleware
+
+
+class Addon:
+    @staticmethod
+    def _check_settings(settings: BaseSettings) -> Optional[str]:
+        if (
+            settings.getwithbase("DOWNLOAD_HANDLERS")["http"]
+            != "scrapy.core.downloader.handlers.http.HTTPDownloadHandler"
+        ):
+            return "'http' value in the 'DOWNLOAD_HANDLERS'"
+        if (
+            settings.getwithbase("DOWNLOAD_HANDLERS")["https"]
+            != "scrapy.core.downloader.handlers.http.HTTPDownloadHandler"
+        ):
+            return "'https' value in the 'DOWNLOAD_HANDLERS'"
+        if (
+            settings.get("REQUEST_FINGERPRINTER_CLASS")
+            != "scrapy.utils.request.RequestFingerprinter"
+        ):
+            return "'REQUEST_FINGERPRINTER_CLASS'"
+        return None
+
+    def update_settings(self, settings: BaseSettings) -> None:
+        nondefault_msg = self._check_settings(settings)
+        if nondefault_msg:
+            raise ValueError(
+                f"The {nondefault_msg} setting is set to a custom value, refusing to override it"
```

Comment:
On a high level, the logic for our handlers is this:

* if request is a Zyte API request, use python-zyte-api client
* otherwise, use the original implementation

Do you think it'd be possible to follow this logic in case of custom download handlers?

I think the same applies for request fingerprints; for non-Zyte-API requests ZYTE_API_FALLBACK_REQUEST_FINGERPRINTER_CLASS is used, which is the default one by default.

Reply:
The approach for http(s) handlers sounds good to me - fallback to default handler, don't inherit. What are your concerns about fragility, what possible issues do you see?

Reply:
My concerns were mostly "what if Scrapy changes internal handling of download handlers, e.g. adds some explicit cleanup for ones in `Downloader.handlers`", but such cleanup already exists: `scrapy.core.downloader.handlers.DownloadHandlers` calls `.close()` for each handler it contains in the `engine_stopped` handler. So we already need to call `.close()` on our internal handler instance manually, and if that part of Scrapy changes then also change something.

Reply:
As a first step for this I changed the existing feature of the hardcoded fallback, it looks like it works, now we need to decide how to pass the user-set handler to our code, as when the handler is loaded it cannot know the old overwritten setting value. I don't like the idea of setting a special setting in the addon code but I'm not sure what other ways we have to pass this.

Reply:
Can we handle it in the same way as request fingerprints? Before changing the handlers, addon reads the current value of the handler settings, and puts it to a setting. Download handler accesses the setting and loads the old, now overwritten, handler as a fallback.

Reply:
That's the solution I said I don't like :)

Reply:
Ok :)

Reply:
> Can we handle it in the same way as request fingerprints? 

Also, if by this you mean `ZYTE_API_FALLBACK_REQUEST_FINGERPRINTER_CLASS` then we currently only tell the user to set it when they want, we don't set it outrselves. Should the addon set it to the old value of `REQUEST_FINGERPRINTER_CLASS`? Or maybe only if it's not set yet?

Reply:
> Also, if by this you mean ZYTE_API_FALLBACK_REQUEST_FINGERPRINTER_CLASS then we currently only tell the user to set it when they want, we don't set it outrselves. Should the addon set it to the old value of REQUEST_FINGERPRINTER_CLASS? 

Yes, that's what I was thinking. 

> Or maybe only if it's not set yet?

Yes, it seems reasonable.

Reply:
We may need to modify Scrapy download handler code after all: it looks like a handler doesn't know which scheme it handles, and so it doesn't know which fallback handler to use.